Reorganize and update documentation in mem_* headers

dreamer opened the pull request:
This is WIP, I still have few changes to make.

@kcgen please double-check if I updated documentation correctly (or if I haven't introduced any problems).

In the meantime, I would like to discuss naming of functions `host_incr*` and `incr_unaligned_*` - each time I read them I am confused and need to remind myself, that those functions exist to add values.

I think the names picked by upstream (e.g. `host_addw`) are better - these functions are almost exclusively used in the context of CPU emulation, thus resemble naming of assembler instructions: `add` and `inc` (in this context, `inc` increases by 1, and does not add a number).

Therefore I think the renaming of these functions would be beneficial:

- `host_incrw` -> `host_addw` (like it was before)
- `incr_unaligned_uint16` -> `add_to_unaligned_uint16`

And, optionally, adding new family of functions:

- `void host_incw(uint8_t *arr)` - increment (by 1) 16-bit LE word pointed by arr
- `void inc_unaligned_int16(uint8_t *arr)` - increment (by 1) 16-bit word pointed by arr

This might actually lead to compilers giving us slightly better code for cases, where previously we would call `incr_unaligned_uint16(ptr, 1)`.

kcgen commented:
Excellent suggestion! This achieves more clarity that the previously proposed `incrementer` functions. 
- `host_add*` and `add_unaligned_*` for arbitrary addition of values to unaligned  
- `host_incr*` and `incr_unaligned_*` for walking the value one 'type size' forward

These patterns play out repeatedly in dynamic cores, and (with the addition of the *_at()` functions, will be more intuitive for to read - especially for the basic unaligned cases where the code previous code performed a type-cast read + type-cast-write + magic-number addtion `+4`, `+2`.

kcgen commented on include/mem_host.h:
```diff
@@ -21,105 +21,185 @@
+ * code such as:
+ *
+ *   #ifdef WORDS_BIGENDIAN
+ *   auto x = byteswap(((uint16_t*)arr)[i]);
+ *   #else
+ *   auto x = ((uint16_t*)arr)[i];
```
Suggest using `[idx]` in the examples to match with the function argument `idx`.

kcgen commented:
The comment changes are all improvements - and I especially like the example block that kicks of each theme of functions.

PS- I had forgot about that branching `#if BIG .. #else ...` scenario that the previous PR fixed too.

dreamer commented:
I updated VS solution, found some typos in the comments and added `add_to_unaligned*` / `inc_unaligned*` / `host_add*` / `host_inc*` functions.

Hopefully, everything is in order, but it would be nice to double-check.

kcgen commented:
We many not end up needing the `+1` incrementers (but no harm in having them for now!)
- `git grep '*(Bit.*++'` is empty
- `git grep '*(Bit.*+='` has hits, but they're not adding 1:

``` text
src/cpu/core_dyn_x86/cache.h:           (*(Bit16u*)&invalidation_map[addr])+=0x101;
src/cpu/core_dyn_x86/cache.h:           (*(Bit32u*)&invalidation_map[addr])+=0x1010101;
src/cpu/core_dyn_x86/cache.h:                   (*(Bit16u*)&invalidation_map[addr])+=0x101;
src/cpu/core_dyn_x86/cache.h:                   (*(Bit32u*)&invalidation_map[addr])+=0x1010101;
src/cpu/core_dyn_x86/decoder.h: *(Bit16u *)&decode.page.wmap[decode.page.index]+=0x0101;
src/cpu/core_dyn_x86/decoder.h: *(Bit32u *)&decode.page.wmap[decode.page.index]+=0x01010101;
src/cpu/core_dyn_x86/decoder.h:         case 2 : (*(Bit16u*)&activecb->cache.wmapmask[mapidx])+=0x0101; break;
src/cpu/core_dyn_x86/decoder.h:         case 4 : (*(Bit32u*)&activecb->cache.wmapmask[mapidx])+=0x01010101; break;
src/cpu/core_dynrec/cache.h:            (*(Bit16u*)&invalidation_map[addr])+=0x101;
src/cpu/core_dynrec/cache.h:            (*(Bit32u*)&invalidation_map[addr])+=0x1010101;
src/cpu/core_dynrec/cache.h:                    (*(Bit16u*)&invalidation_map[addr])+=0x101;
src/cpu/core_dynrec/cache.h:                    (*(Bit32u*)&invalidation_map[addr])+=0x1010101;
src/cpu/core_dynrec/decoder_basic.h:    *(Bit16u *)&decode.page.wmap[decode.page.index]+=0x0101;
src/cpu/core_dynrec/decoder_basic.h:    *(Bit32u *)&decode.page.wmap[decode.page.index]+=0x01010101;
src/cpu/core_dynrec/decoder_basic.h:            case 2 : (*(Bit16u*)&activecb->cache.wmapmask[mapidx])+=0x0101; break;
src/cpu/core_dynrec/decoder_basic.h:            case 4 : (*(Bit32u*)&activecb->cache.wmapmask[mapidx])+=0x01010101; break;
```

Lots of use-cases for the `_at` functions, though:

- `git grep '\*(Bit16u\*).*+2' | wc -l`: 123 hits
- `git grep '\*(Bit32u\*).*+4'` | wc -l: 68 hits
